packages/assets-controllers/src/RatesController/RatesController.ts

```diff
@@ -0,0 +1,187 @@
+import type {
+  RatesMessenger,
+  ConversionRates,
+  RatesControllerState,
+  RatesControllerOptions,
+} from './types';
+
+export const name = 'RatesController';
+
+export enum Cryptocurrency {
+  btc = 'btc',
+}
+
+const DEFAULT_INTERVAL = 180000;
+
+const metadata = {
+  currency: { persist: true, anonymous: true },
+  rates: { persist: true, anonymous: true },
+  fromCurrencies: { persist: true, anonymous: true },
+};
+
+const defaultState = {
+  currency: 'usd',
+  rates: {
+    [Cryptocurrency.btc]: {
+      conversionDate: 0,
+      conversionRate: '0',
+      usdConversionRate: null,
+    },
+  },
+  fromCurrencies: [Cryptocurrency.btc],
+};
+
+export class RatesController extends BaseController<
+  typeof name,
+  RatesControllerState,
+  RatesMessenger
+> {
+  readonly #mutex = new Mutex();
+
+  readonly #fetchMultiExchangeRate;
+
+  readonly #includeUsdRate;
+
+  #intervalLength: number;
+
+  #intervalId: NodeJS.Timeout | undefined;
+
+  /**
+   * Creates a RatesController instance.
+   *
+   * @param options - Constructor options.
+   * @param options.includeUsdRate - Keep track of the USD rate in addition to the current currency rate.
+   * @param options.interval - The polling interval, in milliseconds.
+   * @param options.messenger - A reference to the messaging system.
+   * @param options.state - Initial state to set on this controller.
+   * @param options.fetchMultiExchangeRate - Fetches the exchange rate from an external API. This option is primarily meant for use in unit tests.
+   */
+  constructor({
+    interval = DEFAULT_INTERVAL,
+    messenger,
+    state,
+    includeUsdRate,
+    fetchMultiExchangeRate = defaultFetchExchangeRate,
+  }: RatesControllerOptions) {
+    super({
+      name,
+      metadata,
+      messenger,
+      state: { ...defaultState, ...state },
+    });
+    this.#includeUsdRate = includeUsdRate;
+    this.#fetchMultiExchangeRate = fetchMultiExchangeRate;
+    this.#intervalLength = interval;
+  }
+
+  async #withLock<R>(f: () => R) {
+    const releaseLock = await this.#mutex.acquire();
+    try {
+      return f();
+    } finally {
+      releaseLock();
+    }
+  }
+
+  /**
+   * Executes the polling operation to update rates.
+   */
+  async #executePoll(): Promise<void> {
+    await this.updateRates();
+  }
+
+  /**
+   * Updates the rates by fetching new data.
+   */
+  async updateRates(): Promise<void> {
+    await this.#withLock(async () => {
+      const { currency, fromCurrencies } = this.state;
+      const response = await this.#fetchMultiExchangeRate(
+        currency,
+        fromCurrencies,
+        this.#includeUsdRate,
+      );
+
+      const updatedRates: ConversionRates = {};
+      for (const [key, value] of Object.entries(response)) {
+        updatedRates[key] = {
+          // Divided by 1000 to convert to ms
+          conversionDate: Date.now() / 1000,
+          conversionRate: value[currency] as Cryptocurrency,
```

Comment:
Should be `as string` here no? Or we could even get rid of the type assertion completely?

Reply:
[9b7125a](https://github.com/MetaMask/core/pull/4242/commits/9b7125a5551f691f73600ff8b4755cd210ee79a5)